java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java

```diff
@@ -21,79 +21,68 @@
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
+
 import org.sonar.java.se.symbolicvalues.RelationalSymbolicValue.Kind;
 
 import javax.annotation.CheckForNull;
 
-import java.util.ArrayList;
 import java.util.Collection;
-import java.util.Collections;
+import java.util.EnumSet;
 import java.util.HashSet;
-import java.util.List;
-import java.util.Objects;
 import java.util.Set;
 
-public abstract class BinaryRelation {
+import static org.sonar.java.se.symbolicvalues.RelationalSymbolicValue.Kind.EQUAL;
+import static org.sonar.java.se.symbolicvalues.RelationalSymbolicValue.Kind.GREATER_THAN_OR_EQUAL;
+import static org.sonar.java.se.symbolicvalues.RelationalSymbolicValue.Kind.LESS_THAN;
+import static org.sonar.java.se.symbolicvalues.RelationalSymbolicValue.Kind.METHOD_EQUALS;
+import static org.sonar.java.se.symbolicvalues.RelationalSymbolicValue.Kind.NOT_EQUAL;
+import static org.sonar.java.se.symbolicvalues.RelationalSymbolicValue.Kind.NOT_METHOD_EQUALS;
+
+public class BinaryRelation {
+
+  private static final int MAX_ITERATIONS = 10_000;
+
+  private static final Set<Kind> NORMALIZED_OPERATORS = EnumSet.of(
+      EQUAL, NOT_EQUAL,
+      LESS_THAN, GREATER_THAN_OR_EQUAL,
+      METHOD_EQUALS, NOT_METHOD_EQUALS);
 
   protected final Kind kind;
   protected final SymbolicValue leftOp;
   protected final SymbolicValue rightOp;
-  protected BinaryRelation symmetric;
-  protected BinaryRelation inverse;
+  private BinaryRelation inverse;
   private final int hashcode;
 
-  protected BinaryRelation(Kind kind, SymbolicValue v1, SymbolicValue v2) {
+  private BinaryRelation(Kind kind, SymbolicValue v1, SymbolicValue v2) {
+    if (!NORMALIZED_OPERATORS.contains(kind)) {
+      throw new IllegalArgumentException("Relation " + v1 + kind.operand + v2 + " not normalized!");
+    }
     this.kind = kind;
     leftOp = v1;
     rightOp = v2;
-    hashcode = Objects.hash(kind, leftOp, rightOp);
+    hashcode = kind.hashCode() + leftOp.hashCode() + rightOp.hashCode();
```

Comment:
This is bound to bite us : operand hashcode is its id. so that means we are more than likely to have collisisions with this approach and not "only" on not commutative operators.
I would sort operands in case of commutative operators (based on id) so you end up with same hashcode for equal relations and you can still rely on Objects.hash.